src/v/archival/archival_policy.cc

```diff
@@ -75,6 +90,74 @@ bool archival_policy::upload_deadline_reached() {
     return _upload_deadline < now;
 }
 
+archival_policy::lookup_result archival_policy::find_compacted_segment(
+  model::offset start_offset, storage::log log) {
+    vlog(
+      archival_log.debug,
+      "Finding next compacted segment for {} , start_offset: {}",
+      _ntp,
+      start_offset);
+
+    auto maybe_plog = get_concrete_log_impl(std::move(log));
+    if (!maybe_plog) {
+        vlog(
+          archival_log.debug,
+          "Finding next compacted segment for {}: can't find candidate, no "
+          "segments or "
+          "in-memory log",
+          _ntp);
+        return {};
+    }
+
+    auto plog = maybe_plog.value();
+    const auto& set = plog->segments();
+    auto it = set.lower_bound(start_offset);
+
+    // iterator can reach end for two reasons, either all segments are behind
+    // start_offset or all segments are ahead of start_offset,
+    // which can happen in case of prefix truncation. In the second case, we
+    // want to find the first segment ahead of the start_offset.
+    if (it == set.end() && start_offset < plog->offsets().start_offset) {
+        it = set.begin();
+    }
+
+    if (it == set.end()) {
+        vlog(
+          archival_log.debug,
+          "Finding next compacted segment for {}: can't find candidate, "
+          "all "
+          "segment offsets are "
+          "less than start_offset: {}",
+          _ntp,
+          start_offset);
+        return {};
+    }
+
+    // it now points to either a real segment containing start_offset or the
+    // first segment ahead of start_offset.
+    const auto& segment = *it;
+
+    // This check may incorrectly fail after a node starts up, because even a
+    // compacted segment will have `finished_self_compaction()` to false by
+    // default. The first invocation of housekeeping scan by log manager will
+    // mark all compacted segments as `finished_self_compaction()`, by checking
+    // `.compaction_index` file state.
+    if (segment->finished_self_compaction()) {
+        return {
+          .segment = segment, .ntp_conf = &plog->config(), .forced = false};
```

Comment:
is it possible to take a lock here? If I understand correctly, the compacted segment might be collectable by the storage gc since and we don't want this to happen before the segment is uploaded

Reply:
we do take a lock quite soon after https://github.com/redpanda-data/redpanda/blob/bd2536982907dc1052c080f6e78b4114256f4717/src/v/archival/ntp_archiver_service.cc#L559 when uploading. 

If that is too late we can add an optional lock to the upload candidate and pass it to the scheduled upload object?